wasm/libs/default_executor/src/multisig/transaction.rs

```diff
@@ -0,0 +1,249 @@
+//! Validation and execution logic of instructions for multisig transactions
+
+use alloc::{
+    collections::{btree_map::BTreeMap, btree_set::BTreeSet},
+    vec::Vec,
+};
+
+use super::*;
+
+impl VisitExecute for MultisigPropose {
+    fn visit(&self, executor: &mut Executor) {
+        let host = executor.host();
+        let proposer = executor.context().authority.clone();
+        let target_account = self.account.clone();
+        let multisig_role = multisig_role_for(&target_account);
+        let instructions_hash = HashOf::new(&self.instructions);
+
+        let Ok(_role_found) = host
+            .query(FindRolesByAccountId::new(proposer))
+            .filter_with(|role_id| role_id.eq(multisig_role))
+            .execute_single()
+        else {
+            deny!(executor, "not qualified to propose multisig");
+        };
+
+        let Err(_proposal_not_found) = host.query_single(FindAccountMetadata::new(
+            target_account.clone(),
+            approvals_key(&instructions_hash),
+        )) else {
```

Comment:
```suggestion
        if !host.query_single(FindAccountMetadata::new(
            target_account.clone(),
            approvals_key(&instructions_hash)).is_ok() {
```

Reply:
strictly speaking there could be concurrent multisig transactions with the same set of instructions. Hashes would then overlap. Should this be a concern?

Reply:
#5224